Improve documentation for critbits

konsumlamm opened the pull request:
* Link the mentioned paper
* Simplify examples
* Add more links
* Fix grammar issues
* Make "See also:" bold
* Format

konsumlamm committed:
Improve documentation for critbits

Link the mentioned paper
Simplify examples
Add more links

juancarlospaco commented:
No need to move `runnableExamples` at the top of the file, docgen will do that for you, AFAIK most of stdlib has it near the bottom.

Do not move `itemsWithPrefix` around, why?.

IMHO if you gonna make a Documentation PR do not touch the code, so is easy to merge and backport. :)

Should  all "see also" be bold? :thinking: Bold is to hint a sentence very important to the meaning of the rest of the text, is like 
# SEE ALSOOO!!1!

konsumlamm commented:
> No need to move `runnableExamples` at the top of the file, docgen will do that for you, AFAIK most of stdlib has it near the bottom.

I moved the `runnableExamples` to the top since it's there in the generated docs, I'm aware that that's done automatically. Most modules I've looked at have them at the top too (e.g. `algorithm`, `deques`, `heapqueue`, `options`, `json`).

> Do not move `itemsWithPrefix` around, why?.

I moved it for consistency with the order of the normal iterators (`keys`, `values`, `mvalues`, `items`, `pairs`, `mpairs`).

> IMHO if you gonna make a Documentation PR do not touch the code, so is easy to merge and backport. :)

The only code I changed were runnableExamples, which I'd consider part of the documentation (and moving around `itemsWithPrefix`).

> Should all "see also" be bold? thinking Bold is to hint a sentence very important to the meaning of the rest of the text, is like
> # SEE ALSOOO!!1!

Examples (generated from `runnableExamples`) are also bold, which i think have a similar status, so making "See also:" bold too is more consistent IMO. Then you also immediately see similar procs etc., which is useful when you're searching for something specific.

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -333,10 +353,11 @@ iterator keys*[T](c: CritBitTree[T]): string =
-    c["key1"] = 1
-    c["key2"] = 2
+    let c = {"key1": 1, "key2": 2}.toCritBitTree
     var vals: seq[int]
```
use sugar.collect to simplify test
(plus wherever relevant)

konsumlamm replied:
For clarification, you mean for this?
```nim
var vals: seq[int]
for val in c.values:
  vals.add(val)
```
In that case, i think using `sequtils.toSeq` is even better.

timotheecour replied:
indeed, toSeq is good enough in this case; in other cases, if needed, use collect

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -401,29 +423,14 @@ proc allprefixedAux[T](c: CritBitTree[T], key: string;
   ## Yields all keys starting with `prefix`.
+  ##
+  ## If `longestMatch` is true and no key starts with `prefix`, only
+  ## the longest match is yielded, it doesn't have to be a complete match then.
```
can you add an example for longestMatch=true?

konsumlamm replied:
Hmm, now I'm actually confused how `longestMatch` is supposed to work:
```nim
let c = {"key1": 1, "key2": 2}.toCritBitTree
echo toSeq(c.keysWithPrefix("lol", longestMatch = true))  # => @["key1", "key2"]
echo toSeq(c.keysWithPrefix("keys", longestMatch = true)) # => @["key2"]
```
Apparently it doesn't always yield every key with a longest match, but it doesn't always only yield a single element either...

timotheecour replied:
looks super buggy, can you please open an issue to track this? and in your PR, just skip the runnableExamples for now until we figure out what to do here

konsumlamm replied:
Done: #16575.

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -453,19 +464,39 @@ iterator mvaluesWithPrefix*[T](c: var CritBitTree[T], prefix: string;
+iterator itemsWithPrefix*[T](c: CritBitTree[T], prefix: string;
+                             longestMatch = false): string =
+  ## Alias for `keysWithPrefix <#keysWithPrefix.i,CritBitTree[T],string>`_.
+  runnableExamples:
```
skip this `runnableExamples:`, it adds 0 value since it's an alias

(ditto in all similar cases)

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -453,19 +464,39 @@ iterator mvaluesWithPrefix*[T](c: var CritBitTree[T], prefix: string;
                              longestMatch = false): tuple[key: string, val: T] =
   ## Yields all (key, value)-pairs of `c` starting with `prefix`.
+  ##
+  ## If `longestMatch` is true and no key starts with `prefix`, only
```
even doc comments should be DRY; use this instead:
```nim
`longestMatch`: see `keysWithPrefix`.
```
(ditto elsewhere)

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -476,20 +507,23 @@ iterator pairsWithPrefix*[T](c: CritBitTree[T],
-  ## ``{keyA: value, keyB: value}``, ``{:}``
-  ## If `T` is void the outputs look like:
-  ## ``{keyA, keyB}``, ``{}``.
+  ## `{keyA: value, keyB: value}`, `{:}`.
```
replace this verbose explanation by a runnableExamples; you can even skip the whole doc comment, it adds 0 value once you have a runnableExamples (`$` is standard)

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -8,10 +8,34 @@
+## [excellent paper by Adam Langley](https://www.imperialviolet.org/binary/critbit.pdf).
 ## (A crit bit tree is a form of `radix tree`:idx: or `patricia trie`:idx:.)
 
+runnableExamples:
```
I know it's pre-existing but it's pretty bad for a top-level runnableExamples that's supposed to explain the main points. 

Any improvements welcome; eg using `toCritBitTree` etc; nothing should be redundant here (tests are there for that)

konsumlamm committed:
Improve critbits examples

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -342,10 +333,11 @@ iterator leaves[T](n: Node[T]): Node[T] =
-    var keys: seq[string]
-    for key in c.keys:
-      keys.add(key)
+    from sequtils import toSeq
+
+    let
+      c = {"key1": 1, "key2": 2}.toCritBitTree
+      keys = toSeq(c.keys)
     doAssert keys == @["key1", "key2"]
```
avoid variables used just once, prefer concise runnableExamples, eg:
```nim
from sequtils import toSeq
let c = {"key1": 1, "key2": 2}.toCritBitTree
doAssert toSeq(c.keys) == @["key1", "key2"]
```

(main difference is avoiding intermediate `keys` variable)

ditto elsewhere

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -511,19 +482,27 @@ iterator mpairsWithPrefix*[T](c: var CritBitTree[T],
-  ## If `T` is `void`, the outputs look like:
-  ## `{keyA, keyB}`, `{}`.
+  ## Turns `c` into a string representation.
+  runnableExamples:
+    block:
+      var c: CritBitTree[int]
+      assert $c == "{:}"
+      c["key1"] = 1
+      c["key2"] = 2
+      assert $c == """{"key1": 1, "key2": 2}"""
+    block:
+      var c: CritBitTree[void]
+      assert $c == "{}"
+      c.incl("key1")
+      c.incl("key2")
+      assert $c == """{"key1", "key2"}"""
```
prefer concision:
```nim
  runnableExamples:
    doAssert $CritBitTree[int].default == "{:}"
    doAssert ${"key1": 1, "key2": 2}.toCritBitTree == """{"key1": 1, "key2": 2}"""
    doAssert $CritBitTree[void].default == "{}"
    doAssert $["key1", "key2"].toCritBitTree == """{"key1", "key2"}"""
```

timotheecour commented on lib/pure/collections/critbits.nim:
```diff
@@ -13,28 +13,24 @@
+  critbitAsSet.excl("")
+  doAssert "" notin critbitAsSet
+  doAssert toSeq(critbitAsSet.items) == @["kitten", "puppy"]
+
```
add:
```nim
  let same = ["puppy", "kitten", "puppy"].toCritBitTree
  doAssert toSeq(same.keys) == toSeq(critbitAsSet.keys)
```

timotheecour reviewed (approved):
LGTM after addressing last comments

konsumlamm committed:
Make examples more concise

timotheecour commented:
@konsumlamm thanks! similar PR's welcome, in particular ones that change code-blocks to runnableExamples (refs: https://github.com/nim-lang/RFCs/issues/309#issuecomment-752821157)
